sympy/parsing/sympy_parser.py

```diff
@@ -953,17 +958,50 @@ def parse_expr(s, local_dict=None, transformations=standard_transformations,
 
     if local_dict is None:
         local_dict = {}
+    elif not isinstance(local_dict, dict):
+        raise TypeError('expecting local_dict to be a dict')
 
     if global_dict is None:
         global_dict = {}
         exec_('from sympy import *', global_dict)
-
+    elif not isinstance(global_dict, dict):
+        raise TypeError('expecting global_dict to be a dict')
+
+    transformations = transformations or ()
+    if transformations:
+        if not iterable(transformations):
+            raise TypeError(
+                '`transformations` should be a list of functions.')
+        for _ in transformations:
+            if not callable(_):
+                raise TypeError(filldedent('''
+                    expected a function in transformations,
+                    not %s''' % func_name(_)))
+            if arity(_) != 3:
+                raise TypeError(filldedent('''
+                    transformations should be functions that
+                    takes 3 arguments.'''))
     code = stringify_expr(s, local_dict, global_dict, transformations)
 
     if not evaluate:
         code = compile(evaluateFalse(code), '<string>', 'eval')
 
-    return eval_expr(code, local_dict, global_dict)
+    rv = eval_expr(code, local_dict, global_dict)
+    def check(e):
+        if type(e) in (list, tuple, set):
+            for i in e:
+                check(i)
+        elif isinstance(e, Basic):
+            for i in e.atoms(Symbol):
+                if i.name[0] in string.digits and \
+                        S(i.name).is_Number:
```

Comment:
Would this double check necessary if we can check that the first character is a digit?

Reply:
I made this more robust by checking to see that it is a valid name since there might be other unexpected ways to mis-parse an expression.